src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```
@@ -1807,6 +1877,160 @@ public override TypeKind TypeKind
+                    {
+                        if (foundMarkerMethod != null)
+                        {
+                            return false;
+                        }
+
+                        foundMarkerMethod = methodHandle;
+                    }
+                }
+
+                return foundMarkerMethod is not null;
+            }
+            catch (BadImageFormatException)
+            {
+            }
+
+            return false;
+        }
+
+        private void DecodeExtensionType(out TypeSymbol underlyingType, out ImmutableArray<NamedTypeSymbol> baseExtensions)
+        {
+            // PROTOTYPE we should also decode `implicit` vs. `explicit`
+            // PROTOTYPE consider optimizing/caching
+
+            TypeSymbol? foundUnderlyingType;
+            if (!tryDecodeExtensionType(out foundUnderlyingType, out baseExtensions))
+            {
+                if (foundUnderlyingType is null)
+                {
+                    foundUnderlyingType = new UnsupportedMetadataTypeSymbol();
+                }
+
+                if (baseExtensions.IsDefault)
+                {
+                    baseExtensions = ImmutableArray<NamedTypeSymbol>.Empty;
+                }
+            }
+
+            underlyingType = foundUnderlyingType;
+
+            return;
+
+            bool tryDecodeExtensionType([NotNullWhen(true)] out TypeSymbol? underlyingType, out ImmutableArray<NamedTypeSymbol> baseExtensions)
+            {
+                var found = IsExtensionType(out MethodDefinitionHandle? foundMarkerMethod);
+                Debug.Assert(found);
+                var markerMethod = foundMarkerMethod!.Value;
+                var moduleSymbol = this.ContainingPEModule;
+
+                underlyingType = null;
+                baseExtensions = default;
+
+                MethodAttributes localFlags;
+                moduleSymbol.Module.GetMethodDefPropsOrThrow(markerMethod, name: out _, implFlags: out _, out localFlags, rva: out _);
```

>GetMethodDefPropsOrThrow [](http://example.com/codeflow?start=36&length=24)

Exceptions are not handled.